File: paper_artifacts/experiments/llr9/artifacts/sources/oss120b-fortran/wf_diff_skew__submitted.f90
! wf_diff_skew.f90
! Fortran implementation of the wf_diff_skew kernel for the OptArena benchmark.
! The driver expects a C‑linkage entry point with the suffix "_fp64" and the
! signature:
!   void wf_diff_skew_fp64(double *a, int64_t len_2d, uint8_t *workspace,
!                         int64_t workspace_bytes);
! The workspace arguments are unused for this kernel but are required by the
! harness.  The array "a" is a square double‑precision matrix stored in
! column‑major order (Fortran layout) with dimensions (len_2d, len_2d).
!
! The kernel computes:
!   a(i,j) = a(i,j) + a(i-1,j) + a(i-1,j+1)
! for i = 2 .. len_2d and j = 1 .. len_2d‑1 (1‑based indexing).  This matches
! the reference implementation in /shared/tasks/wf_diff_skew/wf_diff_skew_numpy.py.
!
subroutine wf_diff_skew_fp64(a, len_2d, workspace, workspace_bytes) bind(C, name="wf_diff_skew_fp64")
    use iso_c_binding
    implicit none

    ! Arguments coming from the harness
    type(c_ptr), value :: a               ! pointer to the matrix data (double*)
    integer(c_int64_t), value :: len_2d   ! matrix dimension
    type(c_ptr), value :: workspace       ! not used, required by the API
    integer(c_int64_t), value :: workspace_bytes

    ! Fortran view of the data
    real(c_double), pointer :: arr(:)
    integer(c_int) :: n, i, j, idx, size

    ! Convert the 64‑bit length to the 32‑bit integer used for loop counters.
    n = int(len_2d, kind=c_int)

    ! Map the raw pointer to a 1‑D Fortran array.
    size = n * n
    call c_f_pointer(a, arr, [size])

    do i = 2, n
        do j = 1, n-1
            idx = (i-1) * n + j
            arr(idx) = arr(idx) + arr(idx - n) + arr(idx - n + 1)
        end do
    end do

end subroutine wf_diff_skew_fp64

File: paper_artifacts/experiments/llr9/artifacts/sources/oss120b-c-skills/wf_diff_skew__submitted.c
#include <stdint.h>
#include <omp.h>

/*
 * Wavefront difference kernel with skew.
 * In-place update of a 2D square array a (row-major layout).
 * a[i][j] = a[i][j] + a[i-1][j] + a[i-1][j+1]
 * for i = 1..LEN_2D-1, j = 0..LEN_2D-2.
 *
 * The reference uses double precision; the engine expects the entry point named
 * `wf_diff_skew_fp64` with the standard signature:
 *   void wf_diff_skew_fp64(double *restrict a,
 *                         int64_t LEN_2D,
 *                         uint8_t *restrict workspace,
 *                         int64_t workspace_bytes);
 * The workspace arguments are unused for this kernel but must be present.
 */

void wf_diff_skew_fp64(double *restrict a, int64_t LEN_2D, uint8_t *restrict workspace, int64_t workspace_bytes) {
    (void)workspace;        // Unused, silence compiler warnings.
    (void)workspace_bytes;  // Unused.

    // Parallelize the inner column loop; outer row loop is inherently sequential due
    // to the dependence on the previous row.
    #pragma omp parallel
    {
        for (int64_t i = 1; i < LEN_2D; ++i) {
            double *restrict a_i   = a + i * LEN_2D;
            double *restrict a_im1 = a + (i - 1) * LEN_2D;
            #pragma omp for simd schedule(static)
            for (int64_t j = 0; j < LEN_2D - 1; ++j) {
                a_i[j] = a_i[j] + a_im1[j] + a_im1[j + 1];
            }
        }
    }
}

